source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +117,185 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
     return;
   }
 
-  // TODO(abaptiste): Resolve the requested name
+  // Resolve the requested name
+  const auto response = getResponseForQuery(query_context);
+
+  // We were not able to satisfy the request locally. Return an empty response to the client
+  if (response == DnsLookupResponseCode::Failure) {
+    sendDnsResponse(std::move(query_context));
+    return;
+  }
+
+  // TODO(abaptiste): external resolution
 
-  // Send an answer to the client
+  // We have an answer. Send it to the client
   sendDnsResponse(std::move(query_context));
 }
 
 void DnsFilter::sendDnsResponse(DnsQueryContextPtr query_context) {
   Buffer::OwnedImpl response;
-  // TODO(abaptiste): serialize and return a response to the client
 
+  // Serializes the generated response to the parsed query from the client. If there is a
+  // parsing error or the incoming query is invalid, we will still generate a valid DNS response
+  message_parser_.buildResponseBuffer(query_context, response);
   Network::UdpSendData response_data{query_context->local_->ip(), *(query_context->peer_),
                                      response};
   listener_.send(response_data);
 }
 
+DnsLookupResponseCode DnsFilter::getResponseForQuery(DnsQueryContextPtr& context) {
+  /* It appears to be a rare case where we would have more than one query in a single request.
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
+    // always attempt to resolve with the configured domains
+    if (isKnownDomain(query->name_) || !config_->forwardQueries()) {
+      // Determine whether the name is a cluster. Move on to the next query if successful
+      if (resolveViaClusters(context, *query)) {
+        continue;
+      }
+
+      // Determine whether we an answer this query with the static configuration
+      if (resolveViaConfiguredHosts(context, *query)) {
+        continue;
+      }
+    }
+    // TODO(abaptiste): resolve the query externally
+  }
+
+  if (context->answers_.empty()) {
+    return DnsLookupResponseCode::Failure;
+  }
+  return DnsLookupResponseCode::Success;
+}
+
+uint32_t DnsFilter::getDomainTTL(const absl::string_view domain) {
```

Comment:
nit: return std::chrono and only convert at the very end.

Reply:
will update